File: ddz/src/app/formula/cardFormula.lua
local consts = require('app.consts')

table.mapnew = function(t, fn)
    local newt = {}
    for k,v in pairs(t) do
        newt[k] = fn(v)
    end
    return newt
end

string.splitEx = function(input, delimiter)
    local ret = {}
    if delimiter == '' then
        for i=1,string.len(input) do
            ret[#ret+1] = string.sub(input, i, i)
        end
        return ret
    end
    return string.split(input, delimiter)
end

local ifTrueThen = function(cond, this, that)
    if cond then return this else return that end
end

local exp = {}

exp.unpackCards = function(cards)
    local unpacked = {}
    for i,v in ipairs(cards) do
        local point = string.sub(v, 2)
        if point == 'X' or point == 'Y' then return {consts.card.suit.joker, point} end
        if point == '0' then point = '10' end
        unpacked[i] = {tonumber(string.sub(v, 1, 1)), point}
    end
    return unpacked
end

exp.unpackCard = function(card)
    local point = string.sub(card, 2)
    if point == '0' then point = '10' end
    return {tonumber(string.sub(card, 1, 1)), point}
end

exp.packCard = function(suit, point)
    if point == 'X' then return point end
    if point == 'Y' then return point end
    if point == '10' then point = '0' end
    return suit .. point
end

exp._toPointIdx = function(c, isPoint)
    local p = ifTrueThen(isPoint, c, string.sub(c, 2)) --c.substring(1);
    if p == consts.card.joker then
        return #consts.card.points + 1;
    elseif p == consts.card.jokerRed then
        return #consts.card.points + 2;
    else
        return table.keyof(consts.card.points, p) -- _.indexOf(consts.card.points, p);
    end
end

exp.sortCards = function(cards, isPoint)
    table.sort(cards, function(c1, c2)
        local p1 = exp._toPointIdx(c1, isPoint);
        local p2 = exp._toPointIdx(c2, isPoint);
        if p1 == p2 and not isPoint then return tonumber(string.sub(c1, 1, 1)) < tonumber(string.sub(c2, 1, 1)) end
        return p1 < p2
    end);
    return cards
end

exp.isCardsValid = function(cards, isPoint, sorted)
    local res = exp.getHandType(cards, isPoint, sorted) ~= -1;
    --printInfo('exp.isCardsValid: cards=%s, isPoint=%s, sorted=%s, res=%s', json.encode(cards), isPoint, sorted, res);
    return res;
end

exp.isCardsGreater = function(cards, compared, isPoint, sorted)
    --printInfo('cards: %s, compared: %s', json.encode(cards), json.encode(compared));
    local handInfo = exp.getHandTypeInfo(cards, isPoint, sorted);
    local compHandInfo = exp.getHandTypeInfo(compared, isPoint, sorted);
    if not handInfo or not compHandInfo then return false end
    --printInfo('handInfo: %s, compHandInfo: %s', json.encode(handInfo), json.encode(compHandInfo));
    if handInfo.type == consts.card.handTypes.ROCKET then
        return true;
    end
    if compHandInfo.type == consts.card.handTypes.ROCKET then
        return false;
    end
    if handInfo.type == consts.card.handTypes.BOMB and
        compHandInfo.type ~= consts.card.handTypes.BOMB then
        return true;
    end
    if handInfo.type ~= compHandInfo.type then
        return false;
    end
    if handInfo.info.required and
        handInfo.info.required ~= compHandInfo.info.required then
        return false;
    end
    return handInfo.info.main > compHandInfo.info.main;
end

local handTypes = table.keys(consts.card.handTypes);
local handFuncNames = table.mapnew(handTypes, function(t)
        return 'isHand' .. table.concat(table.mapnew(string.splitEx(string.lower(t), '_'), function(s)
            return string.upper(string.sub(s, 1, 1)) .. string.sub(s, 2)
        end))
    end);
exp.getHandType = function(cards, isPoint, sorted)
    for i,v in pairs(handTypes) do
        if exp[handFuncNames[i]](cards, isPoint, sorted) then
            return consts.card.handTypes[handTypes[i]];
        end
    end
    return -1;
end

exp.getHandTypeInfo = function(cards, isPoint, sorted)
    for i,v in pairs(handTypes) do
        local info = exp[handFuncNames[i]](cards, isPoint, sorted);
        if info then
            return {type = consts.card.handTypes[handTypes[i]], info = info};
        end
    end
    return nil;
end

exp.isHandRocket = function(cards, isPoint, sorted)
    if #cards ~= 2 then
        return false;
    end
    local ps = ifTrueThen(isPoint, cards, table.mapnew(cards, function(c) return string.sub(c, 2) end))
    if (ps[1] == consts.card.joker and ps[2] == consts.card.jokerRed) or
        (ps[2] == consts.card.joker and ps[1] == consts.card.jokerRed) then
        return {main = exp._toPointIdx(cards[#cards], isPoint)};
    end
    return false;
end

exp.isHandBomb = function(cards, isPoint, sorted)
    if #cards ~= 4 then
        return false;
    end
    local ps = ifTrueThen(isPoint, cards, table.mapnew(cards, function(c) return string.sub(c, 2) end));
    if ps[2] == ps[1] and ps[3] == ps[1] and ps[4] == ps[1] then
        return {main = exp._toPointIdx(ps[1], true)};
    end
    return false;
end

exp.isHandSolo = function(cards, isPoint, sorted)
    if #cards ~= 1 then
        return false;
    end
    return {main = exp._toPointIdx(cards[1], isPoint)};
end

exp.isHandPair = function(cards, isPoint, sorted)
    if #cards ~= 2 then
        return false;
    end
    local ps = ifTrueThen(isPoint, cards, table.mapnew(cards, function(c) return string.sub(c, 2) end));
    if ps[2] == ps[1] then
        return {main = exp._toPointIdx(ps[1], true)};
    end
    return false;
end

exp.isHandStraight = function(cards, isPoint, sorted)
    if #cards < 5 then
        return false;
    end
    local ps = ifTrueThen(isPoint, cards, table.mapnew(cards, function(c) return string.sub(c, 2) end));
    ps = ifTrueThen(sorted, ps, exp.sortCards(ps, true));
    local idxs = table.mapnew(ps, function(p) return exp._toPointIdx(p, true) end);
    for i=2,#idxs do
        if idxs[i] ~= idxs[i-1] + 1 then
            return false;
        end
    end
    if idxs[#idxs] >= #consts.card.points then
        return false;
    end
    return {main = idxs[#idxs], required = #cards};
end



exp.isHandConsecutivePairs = function(cards, isPoint, sorted)
    if #cards < 6 or math.fmod(#cards, 2) ~= 0 then
        return false;
    end
    local ps = ifTrueThen(isPoint, cards, table.mapnew(cards, function(c) return string.sub(c, 2) end))
    ps = ifTrueThen(sorted, ps, exp.sortCards(ps, true));
    if table.keyof({'2', consts.card.joker, consts.card.jokerRed}, ps[#ps]) ~= nil then
        return false;
    end
    for i=0,#ps/2-1 do
        if ps[i*2+1] ~= ps[i*2+2] then
            return false;
        end
        if i > 0 and exp._toPointIdx(ps[i*2], true) + 1 ~= exp._toPointIdx(ps[i*2+1], true) then
            return false;
        end
    end
    return {main = exp._toPointIdx(ps[#ps], true), required = #cards/2};
end

exp.isHandTrio = function(cards, isPoint, sorted)
    if #cards ~= 3 then
        return false;
    end
    local ps = ifTrueThen(isPoint, cards, table.mapnew(cards, function(c) return string.sub(c, 2) end))
    if ps[2] == ps[1] and ps[3] == ps[1] then
        return {main = exp._toPointIdx(ps[1], true)};
    end
    return false;
end

exp.isHandTrioSolo = function(cards, isPoint, sorted)
    if #cards ~= 4 then
        return false;
    end
    local ps = ifTrueThen(isPoint, cards, table.mapnew(cards, function(c) return string.sub(c, 2) end))
    printInfo(json.encode(ps))
    ps = ifTrueThen(sorted, ps, exp.sortCards(ps, true));
    if ifTrueThen(ps[2] == ps[1], (ps[3] == ps[1] and ps[4] ~= ps[1]), (ps[3] == ps[2] and ps[4] == ps[2])) then
        return {main = exp._toPointIdx(ps[2], true)};
    end
    return false;
end


exp.isHandTrioPair = function(cards, isPoint, sorted)
    if #cards ~= 5 then
        return false;
    end
    local ps = ifTrueThen(isPoint, cards, table.mapnew(cards, function(c) return string.sub(c, 2) end))
    ps = ifTrueThen(sorted, ps, exp.sortCards(ps, true));
    if ifTrueThen(ps[3] == ps[2], (ps[1] == ps[2] and ps[5] == ps[4] and ps[4] ~= ps[3]),
        (ps[1] == ps[2] and ps[4] == ps[3] and ps[5] == ps[3])) then
        return {main = exp._toPointIdx(ps[3], true)};
    end
    return false;
end

exp.isHandAirplane = function(cards, isPoint, sorted)
    if #cards < 6 or math.fmod(#cards, 3) ~= 0 then
        return false;
    end
    local ps = ifTrueThen(isPoint, cards, table.mapnew(cards, function(c) return string.sub(c, 2) end))
    ps = ifTrueThen(sorted, ps, exp.sortCards(ps, true));
    for i=0,#ps/3-1 do
        if ps[i*3+2+1] ~= ps[i*3+1] or ps[i*3+1+1] ~= ps[i*3+1] then
            return false;
        end
        if i > 0 and exp._toPointIdx(ps[i*3-1+1], true) + 1 ~= exp._toPointIdx(ps[i*3+1], true) then
            return false;
        end
    end
    if ps[#ps] == '2' then
        return false;
    end
    return {main = exp._toPointIdx(ps[#ps], true), required = #cards/3};
end

exp.isHandAirplaneSolo = function(cards, isPoint, sorted)
    if #cards < 8 or math.fmod(#cards, 4) ~= 0 then
        return false;
    end
    local ps = ifTrueThen(isPoint, cards, table.mapnew(cards, function(c) return string.sub(c, 2) end))
    ps = ifTrueThen(sorted, ps, exp.sortCards(ps, true));
    local trios = {};
    for i=3,#ps do
        if ps[i] == ps[i-2] and ps[i-1] == ps[i-2] then
            if table.keyof(trios, ps[i]) == nil then
                trios[#trios+1] = ps[i]
            end
        end
    end
    if #trios < 2 then
        return false;
    end
    if trios[#trios] == '2' then
        return false;
    end
    local trioIdxs = table.mapnew(trios, function(p) return exp._toPointIdx(p, true) end);
    for i=2,#trioIdxs do
        if trioIdxs[i] ~= trioIdxs[i-1] + 1 then
            return false;
        end
    end
    return {main = trioIdxs[#trioIdxs], required = #cards/4};
end

exp.isHandAirplanePair = function(cards, isPoint, sorted)
    if #cards < 10 or math.fmod(#cards, 5) ~= 0 then
        return false;
    end
    local ps = ifTrueThen(isPoint, cards, table.mapnew(cards, function(c) return string.sub(c, 2) end))
    ps = ifTrueThen(sorted, ps, exp.sortCards(ps, true));
    local trios = {};
    for i=3,#ps do
        if ps[i] == ps[i-2] and ps[i-1] == ps[i-2] then
            if table.keyof(trios, ps[i]) == nil then
                trios[#trios+1] = ps[i]
            end
        end
    end
    if #trios < 2 then
        return false;
    end
    if trios[#trios] == '2' then
        return false;
    end
    local trioIdxs = table.mapnew(trios, function(p) return exp._toPointIdx(p, true) end);
    for i=2,#trioIdxs do
        if trioIdxs[i] ~= trioIdxs[i-1] + 1 then
            return false;
        end
    end
    local pairs = {};
    for i=1,#ps do
        if table.keyof(trios, ps[i]) == nil then
            pairs[#pairs+1] = ps[i]
        end
    end
    if exp._isPairs(pairs) then
        return {main = trioIdxs[#trioIdxs], required = #cards/5};
    end
    return false;
end

exp.isHandSpaceShuttleSolo = function(cards, isPoint, sorted)
    if #cards ~= 6 then
        return false;
    end
    local ps = ifTrueThen(isPoint, cards, table.mapnew(cards, function(c) return string.sub(c, 2) end))
    ps = ifTrueThen(sorted, ps, exp.sortCards(ps, true));
    local fours = {};
    for i=4,#ps do
        if ps[i] == ps[i-3] and ps[i-1] == ps[i-3] and ps[i-2] == ps[i-3] then
            if table.keyof(fours, ps[i]) == nil then
                fours[#fours+1] = ps[i]
            end
        end
    end
    if #fours < 1 then
        return false;
    end
    return {main = exp._toPointIdx(fours[#fours], true)};
end

exp._isPairs = function(pairs)
    if math.fmod(#pairs, 2) ~= 0 then
        return false;
    end
    for i=0,#pairs/2-1 do
        if pairs[i*2+1] ~= pairs[i*2+2] then
            return false;
        end
    end
    return true;
end

exp.isHandSpaceShuttlePair = function(cards, isPoint, sorted)
    if #cards ~=8 then
        return false;
    end
    local ps = ifTrueThen(isPoint, cards, table.mapnew(cards, function(c) return string.sub(c, 2) end))
    ps = ifTrueThen(sorted, ps, exp.sortCards(ps, true));
    local fours = {};
    for i=4,#ps do
        if ps[i] == ps[i-3] and ps[i-1] == ps[i-3] and ps[i-2] == ps[i-3] then
            if table.keyof(fours, ps[i]) == nil then
                fours[#fours+1] = ps[i]
            end
        end
    end
    if #fours < 1 then
        return false;
    end
    local pairs = {};
    for i=1,#ps do
        if table.keyof(fours, ps[i]) == nil then
            pairs[#pairs+1] = ps[i]
        end
    end
    if exp._isPairs(pairs) then
        return {main = exp._toPointIdx(fours[#fours], true)};
    end
    return false;
end


return exp

File: server/src/app/formula/cardFormula.js
'use strict';

var _ = require('lodash');
var consts = require('../consts');
var logger = require('pomelo-logger').getLogger('area', __filename);

var exp = module.exports;

exp.cardPoints = {};
exp.cardPoints[consts.card.jokerRed] = consts.card.points.length + 1;
exp.cardPoints[consts.card.joker] = consts.card.points.length;
consts.card.points.forEach(function(k, i){
	exp.cardPoints[k] = i;
});

exp.pointIdx2Point = function(i) {
	if (i === consts.card.points.length) {
		return consts.card.joker;
	} else if (i === consts.card.points.length + 1) {
		return consts.card.jokerRed;
	} else {
		return consts.card.points[i];
	}
};

exp._toPointIdx = function(c, isPoint) {
	var p = isPoint ? c : c.substring(1);
	return exp.cardPoints[p];
};

exp.sortCards = function(cards, isPoint) {
	return _.sortBy(cards, function(c){
		return exp._toPointIdx(c, isPoint);
	});
};

exp.autoPlay = function(cards, must) {
	return must ? [cards[0]] : [];
};

exp.isCardsValid = function(cards, isPoint, sorted) {
	var res = exp.getHandType(cards, isPoint, sorted) !== -1;
	//logger.debug('cardFormula.isCardsValid: cards=%j, isPoint=%s, sorted=%s, res=%s', cards, isPoint, sorted, res);
	return res;
};

exp.isCardsGreater = function(cards, compared, isPoint, sorted) {
	console.log('cards: %j, compared: %j', cards, compared);
	var handInfo = exp.getHandTypeInfo(cards, isPoint, sorted);
	var compHandInfo = exp.getHandTypeInfo(compared, isPoint, sorted);
	//logger.debug('handInfo: %j, compHandInfo: %j', handInfo, compHandInfo);
	if(handInfo.type === consts.card.handTypes.ROCKET) {
		return true;
	}
	if(compHandInfo.type === consts.card.handTypes.ROCKET) {
		return false;
	}
	if(handInfo.type === consts.card.handTypes.BOMB &&
		compHandInfo.type !== consts.card.handTypes.BOMB) {
		return true;
	}
	if(handInfo.type !== compHandInfo.type) {
		return false;
	}
	if(handInfo.info.required &&
		handInfo.info.required !== compHandInfo.info.required) {
		return false;
	}
 	return handInfo.info.main > compHandInfo.info.main;
};

var handTypes = _.keys(consts.card.handTypes);
var handFuncNames = handTypes.map((t) => 'isHand' + t.toLowerCase().split('_').map((s)=>s.charAt(0).toUpperCase() + s.slice(1)).join(''));
exp.getHandType = function(cards, isPoint, sorted) {
	for (var i = 0; i < handTypes.length; i++) {
		if(exp[handFuncNames[i]](cards, isPoint, sorted)) {
			return consts.card.handTypes[handTypes[i]];
		}
	}
	return -1;
};

exp.getHandTypeInfo = function(cards, isPoint, sorted) {
	for (var i = 0; i < handTypes.length; i++) {
		var info = exp[handFuncNames[i]](cards, isPoint, sorted);
		if(info) {
			return {type: consts.card.handTypes[handTypes[i]], info: info};
		}
	}
	return null;
};

exp.isHandRocket = function(cards, isPoint, sorted) {
	if(cards.length !== 2) {
		return false;
	}
	var ps = isPoint ? cards : cards.map((c) => c.substring(1));
	if((ps[0] === consts.card.joker && ps[1] === consts.card.jokerRed) ||
		(ps[1] === consts.card.joker && ps[0] === consts.card.jokerRed)) {
		return {main: exp._toPointIdx(cards[cards.length-1], isPoint)};
	}
	return false;
};

exp.isHandBomb = function(cards, isPoint, sorted) {
	if(cards.length !== 4) {
		return false;
	}
	var ps = isPoint ? cards : cards.map((c) => c.substring(1));
	if(ps[1] === ps[0] && ps[2] === ps[0] && ps[3] === ps[0]) {
		return {main: exp._toPointIdx(ps[0], true)};
	}
	return false;
};

exp.isHandSolo = function(cards, isPoint, sorted) {
	if(cards.length !== 1) {
		return false;
	}
	return {main: exp._toPointIdx(cards[0], isPoint)};
};

exp.isHandPair = function(cards, isPoint, sorted) {
	if(cards.length !== 2) {
		return false;
	}
	var ps = isPoint ? cards : cards.map((c) => c.substring(1));
	if(ps[1] === ps[0]) {
		return {main: exp._toPointIdx(ps[0], true)};
	}
	return false;
};

exp.isHandStraight = function(cards, isPoint, sorted) {
	if(cards.length < 5) {
		return false;
	}
	var ps = isPoint ? cards : cards.map((c) => c.substring(1));
	ps = sorted ? ps : exp.sortCards(ps, true);
	var idxs = ps.map((p) => exp._toPointIdx(p, true));
	for (let i = 1; i < idxs.length; i++) {
		if(idxs[i] !== idxs[i-1] + 1) {
			return false;
		}
	}
	if(idxs[idxs.length-1] >= consts.card.points.length - 1) {
		return false;
	}
	return {main: idxs[idxs.length-1], required: cards.length};
};

exp.isHandConsecutivePairs = function(cards, isPoint, sorted) {
	if(cards.length < 6 || cards.length % 2 !== 0) {
		return false;
	}
	var ps = isPoint ? cards : cards.map((c) => c.substring(1));
	ps = sorted ? ps : exp.sortCards(ps, true);
	if(_.indexOf(['2', consts.card.joker, consts.card.jokerRed], ps[ps.length-1]) !== -1) {
		return false;
	}
	for (var i = 0; i < ps.length / 2; i++) {
		if(ps[i*2] !== ps[i*2 + 1]) {
			return false;
		}
		if(i > 0 && exp._toPointIdx(ps[i*2-1], true) + 1 !== exp._toPointIdx(ps[i*2], true)) {
			return false;
		}
	}
	return {main: exp._toPointIdx(ps[ps.length-1], true), required: cards.length/2};
};

exp.isHandTrio = function(cards, isPoint, sorted) {
	if(cards.length !== 3) {
		return false;
	}
	var ps = isPoint ? cards : cards.map((c) => c.substring(1));
	if(ps[1] === ps[0] && ps[2] === ps[0]) {
		return {main: exp._toPointIdx(ps[0], true)};
	}
	return false;
};

exp.isHandTrioSolo = function(cards, isPoint, sorted) {
	if(cards.length !== 4) {
		return false;
	}
	var ps = isPoint ? cards : cards.map((c) => c.substring(1));
	ps = sorted ? ps : exp.sortCards(ps, true);
	if(ps[1] === ps[0] ? (ps[2] === ps[0] && ps[3] !== ps[0]) : (ps[2] === ps[1] && ps[3] === ps[1])) {
		return {main: exp._toPointIdx(ps[1], true)};
	}
	return false;
};


exp.isHandTrioPair = function(cards, isPoint, sorted) {
	if(cards.length !== 5) {
		return false;
	}
	var ps = isPoint ? cards : cards.map((c) => c.substring(1));
	ps = sorted ? ps : exp.sortCards(ps, true);
	if(ps[2] === ps[1] ? (ps[0] === ps[1] && ps[4] === ps[3] && ps[3] !== ps[2]) :
		(ps[0] === ps[1] && ps[3] === ps[2] && ps[4] === ps[2])) {
		return {main: exp._toPointIdx(ps[2], true)};
	}
	return false;
};

exp.isHandAirplane = function(cards, isPoint, sorted) {
	if(cards.length < 6 || cards.length % 3 !== 0) {
		return false;
	}
	var ps = isPoint ? cards : cards.map((c) => c.substring(1));
	ps = sorted ? ps : exp.sortCards(ps, true);
	for (var i = 0; i < ps.length / 3; i++) {
		if(ps[i*3+2] !== ps[i*3] || ps[i*3+1] !== ps[i*3]) {
			return false;
		}
		if(i > 0 && exp._toPointIdx(ps[i*3-1], true) + 1 !== exp._toPointIdx(ps[i*3], true)) {
			return false;
		}
	}
	if(ps[ps.length -1] === '2') {
		return false;
	}
	return {main: exp._toPointIdx(ps[ps.length-1], true), required: cards.length/3};
};

exp.isHandAirplaneSolo = function(cards, isPoint, sorted) {
	if(cards.length < 8 || cards.length % 4 !== 0) {
		return false;
	}
	var ps = isPoint ? cards : cards.map((c) => c.substring(1));
	ps = sorted ? ps : exp.sortCards(ps, true);
	var trios = [];
	for (let i = 2; i < ps.length; i++) {
		if(ps[i] === ps[i-2] && ps[i-1] === ps[i-2]) {
			if(_.indexOf(trios, ps[i]) === -1) {
				trios.push(ps[i]);
			}
		}
	}
	if(trios.length < 2) {
		return false;
	}
	if(trios[trios.length -1] === '2') {
		return false;
	}
	var trioIdxs = trios.map((p) => exp._toPointIdx(p, true));
	for (let i = 1; i < trioIdxs.length; i++) {
		if(trioIdxs[i] !== trioIdxs[i-1] + 1) {
			return false;
		}
	}
	return {main: trioIdxs[trioIdxs.length-1], required: cards.length/4};
};

exp.isHandAirplanePair = function(cards, isPoint, sorted) {
	if(cards.length < 10 || cards.length % 5 !== 0) {
		return false;
	}
	var ps = isPoint ? cards : cards.map((c) => c.substring(1));
	ps = sorted ? ps : exp.sortCards(ps, true);
	var trios = [];
	for (let i = 2; i < ps.length; i++) {
		if(ps[i] === ps[i-2] && ps[i-1] === ps[i-2]) {
			if(_.indexOf(trios, ps[i]) === -1) {
				trios.push(ps[i]);
			}
		}
	}
	if(trios.length < 2) {
		return false;
	}
	if(trios[trios.length -1] === '2') {
		return false;
	}
	var trioIdxs = trios.map((p) => exp._toPointIdx(p, true));
	for (let i = 1; i < trioIdxs.length; i++) {
		if(trioIdxs[i] !== trioIdxs[i-1] + 1) {
			return false;
		}
	}
	var pairs = [];
	for (let i = 0; i < ps.length; i++) {
		if(_.indexOf(trios, ps[i]) === -1) {
			pairs.push(ps[i]);
		}
	}
	if(exp._isPairs(pairs)) {
		return {main: trioIdxs[trioIdxs.length-1], required: cards.length/5};
	}
	return false;
};

exp.isHandSpaceShuttleSolo = function(cards, isPoint, sorted) {
	if(cards.length !== 6) {
		return false;
	}
	var ps = isPoint ? cards : cards.map((c) => c.substring(1));
	ps = sorted ? ps : exp.sortCards(ps, true);
	var fours = [];
	for (let i = 3; i < ps.length; i++) {
		if(ps[i] === ps[i-3] && ps[i-1] === ps[i-3] && ps[i-2] === ps[i-3]) {
			if(_.indexOf(fours, ps[i]) === -1) {
				fours.push(ps[i]);
			}
		}
	}
	if(fours.length < 1) {
		return false;
	}
	return {main: exp._toPointIdx(fours[fours.length-1], true)};
};

exp._isPairs = function(pairs) {
	if(pairs.length % 2 !== 0) {
		return false;
	}
	for (let i = 0; i < pairs.length/2; i++) {
		if(pairs[i*2] !== pairs[i*2+1]) {
			return false;
		}
	}
	return true;
};

exp.isHandSpaceShuttlePair = function(cards, isPoint, sorted) {
	if(cards.length !== 8) {
		return false;
	}
	var ps = isPoint ? cards : cards.map((c) => c.substring(1));
	ps = sorted ? ps : exp.sortCards(ps, true);
	var fours = [];
	for (let i = 3; i < ps.length; i++) {
		if(ps[i] === ps[i-3] && ps[i-1] === ps[i-3] && ps[i-2] === ps[i-3]) {
			if(_.indexOf(fours, ps[i]) === -1) {
				fours.push(ps[i]);
			}
		}
	}
	if(fours.length < 1) {
		return false;
	}
	var pairs = [];
	for (let i = 0; i < ps.length; i++) {
		if(_.indexOf(fours, ps[i]) === -1) {
			pairs.push(ps[i]);
		}
	}
	if(exp._isPairs(pairs)) {
		return {main: exp._toPointIdx(fours[fours.length-1], true)};
	}
	return false;
};

exp.arrangeCards = function(cards, sorted) {
	var trios = [], solos = [], pairs = [], straights = [], bombs = [];
	cards = sorted ? cards : exp.sortCards(cards, true);
	var ps = cards.map(exp._toPointIdx);
	for (var i = 0; i < ps.length; i++) {
		ps[i]
	}
};


